restaking_program/src/ncn_delegate_token_account.rs

```diff
@@ -0,0 +1,93 @@
+use jito_bytemuck::AccountDeserialize;
+use jito_jsm_core::loader::{
+    load_signer, load_token_2022_program, load_token_account, load_token_mint, load_token_program,
+};
+use jito_restaking_core::ncn::Ncn;
+use solana_program::{
+    account_info::AccountInfo, entrypoint::ProgramResult, msg, program::invoke_signed,
+    program_error::ProgramError, pubkey::Pubkey,
+};
+
+/// Processes the ncn delegate token account instruction: [`crate::RestakingInstruction::NcnDelegateTokenAccount`]
+///
+/// This instruction handles the delegation of tokens from a token account managed by an NCN to a delegate account.
+/// The NCN delegate_admin might call this instruction to delegate authority over tokens managed by the NCN to another account.
+///
+/// # Arguments
+/// * `program_id` - The public key of the program, used to ensure the correct program is being executed.
+/// * `accounts` - A slice of `AccountInfo` representing the accounts required for this instruction.
+/// * `amount` - The number of tokens to delegate to the delegate account.
+///
+/// # Returns
+/// * `ProgramResult` - Returns `Ok(())` if the delegation is successful, otherwise returns an appropriate `ProgramError`.
+pub fn process_ncn_delegate_token_account(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    amount: u64,
+) -> ProgramResult {
+    let [ncn_info, delegate_admin, token_mint, token_account, delegate, token_program_info] =
+        accounts
+    else {
+        return Err(ProgramError::NotEnoughAccountKeys);
+    };
+
+    Ncn::load(program_id, ncn_info, false)?;
+    load_signer(delegate_admin, false)?;
+    load_token_mint(token_mint)?;
+    load_token_account(
+        token_account,
+        ncn_info.key,
+        token_mint.key,
+        token_program_info,
+    )?;
+
+    match (*token_mint.owner, *token_account.owner) {
```

Comment:
This should be cleaned up if we accept both the token program and token extensions program ( token2022 ) in the load_token_program call

Reply:
Then, just have a comparison `token_mint.owner.eq(token_account.owner)` to ensure they match

Reply:
I simplified like this!
```rust
    if token_mint.owner.ne(token_account.owner) {
        return Err(ProgramError::InvalidAccountData);
    }
```